rust/automerge/src/automerge/diff.rs

```diff
@@ -274,171 +273,48 @@ fn get_inc(before: &Winner<'_>, after: &Winner<'_>) -> Option<i64> {
-                    // after   ----------------*
-                    mark(&mut f, after.start, before.start, after);
-                    if after.end > before.start {
-                        after.start = std::cmp::max(after.start, before.start);
-                    } else {
-                        continue 'after_marks;
-                    }
-                }
-                if after.start >= before.start {
-                    // before ------------*
-                    // after    ---------*
-                    if after.start > before.start {
-                        before.start = std::cmp::min(before.start, after.start);
-                        unmark(&mut f, before.start, after.start, before);
-                    }
-                    if after.end > before.end {
-                        if after.data.value != before.data.value {
-                            mark(&mut f, after.start, before.end, after);
-                        }
-                        after.start = before.end;
-                        before.start = before.end;
-                        continue;
-                    } else {
-                        if after.data.value != before.data.value {
-                            mark(&mut f, after.start, after.end, after);
-                        }
-                        before.start = after.end;
-                        continue 'after_marks;
-                    }
-                }
-            }
-            // mark after
-            mark(&mut f, after.start, after.end, after);
-        }
-        for before in &self.before_marks {
-            if before.start != before.end {
-                unmark(&mut f, before.start, before.end, before);
-            }
-        }
-        if !f.is_empty() {
-            Some(f.into_values().flat_map(|v| v.into_iter()).collect())
+    fn current(&self) -> Option<MarkSet> {
+        // do this without all the cloning - cache the result
+        let b = self.before.current().cloned().unwrap_or_default();
+        let a = self.after.current().cloned().unwrap_or_default();
+        if a != b {
+            let result = b.diff(&a);
+            Some(result)
         } else {
             None
         }
     }
-}
 
-fn unmark<'a>(
-    finished: &mut BTreeMap<String, Vec<Mark<'a>>>,
-    start: usize,
-    end: usize,
-    from: &Mark<'a>,
-) {
-    let f_vec = finished.entry(from.data.name.to_string()).or_default();
-    if start < end {
-        if let Some(last) = f_vec.last_mut() {
-            if last.data.name == from.data.name && last.data.value.is_null() && last.end == start {
-                last.end = end;
-                return;
-            }
-        }
-        f_vec.push(Mark {
-            start,
-            end,
-            data: Cow::Owned(MarkData {
-                name: from.data.name.clone(),
-                value: ScalarValue::Null,
-            }),
-        });
+    fn process(&mut self, op: &'a Op) -> Option<Patch<'static>> {
```

Comment:
I believe you can remove the comment above about this being O(n^2) as it is now O(n)?